vm/devices/storage/disk_crypt/block_crypto/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+    }
+
+    pub struct NonStreamingCipherCtx<'a> {
+        ctx: openssl::cipher_ctx::CipherCtx,
+        enc: bool,
+        _dummy: &'a (),
+    }
+
+    pub type Error = openssl::error::ErrorStack;
+
+    pub type AesXts256 = NonStreamingCipher;
+    pub type AesXts256Ctx<'a> = NonStreamingCipherCtx<'a>;
+
+    pub fn aes_xts_256(key: &[u8], _data_unit_size: u32) -> Result<AesXts256, Error> {
+        let mut enc = openssl::cipher_ctx::CipherCtx::new()?;
+        enc.encrypt_init(
+            Some(openssl::cipher::Cipher::aes_256_xts()),
+            Some(key),
+            None,
+        )?;
+        let mut dec = openssl::cipher_ctx::CipherCtx::new()?;
+        dec.decrypt_init(
+            Some(openssl::cipher::Cipher::aes_256_xts()),
+            Some(key),
+            None,
+        )?;
+        Ok(NonStreamingCipher { enc, dec })
+    }
+
+    impl NonStreamingCipher {
+        pub fn ctx(&self, enc: bool) -> Result<NonStreamingCipherCtx<'_>, Error> {
+            let mut ctx = openssl::cipher_ctx::CipherCtx::new()?;
+            ctx.copy(if enc { &self.enc } else { &self.dec })?;
+            Ok(NonStreamingCipherCtx {
+                ctx,
+                enc,
+                _dummy: &(),
+            })
+        }
+    }
+
+    impl NonStreamingCipherCtx<'_> {
+        pub fn cipher(&mut self, iv: &[u8], data: &mut [u8]) -> Result<(), Error> {
+            if self.enc {
+                self.ctx.encrypt_init(None, None, Some(iv))?;
+            } else {
+                self.ctx.decrypt_init(None, None, Some(iv))?;
+            }
+            self.ctx.cipher_update_inplace(data, data.len())?;
+            Ok(())
+        }
+    }
+}
+
+#[cfg(windows)]
+mod bcrypt {
+    // UNSAFETY: calling bcrypt APIs
+    #![allow(unsafe_code)]
+
+    use std::sync::OnceLock;
+    use thiserror::Error;
+    use windows::Win32::Foundation::RtlNtStatusToDosError;
+    use windows::Win32::Foundation::NTSTATUS;
+    use windows::Win32::Security::Cryptography::BCRYPT_ALG_HANDLE;
+    use windows::Win32::Security::Cryptography::BCRYPT_HANDLE;
+    use windows::Win32::Security::Cryptography::BCRYPT_KEY_HANDLE;
+    use windows::Win32::Security::Cryptography::BCRYPT_OPEN_ALGORITHM_PROVIDER_FLAGS;
+
+    #[derive(Debug, Error)]
+    #[error("{op} failed")]
+    pub struct Error {
+        op: &'static str,
+        #[source]
+        err: std::io::Error,
+    }
+
+    pub struct AesXts256(Key);
+
+    pub struct AesXts256Ctx<'a> {
+        key: &'a Key,
+        enc: bool,
+    }
+
+    impl AesXts256 {
+        pub fn ctx(&self, enc: bool) -> Result<AesXts256Ctx<'_>, Error> {
+            Ok(AesXts256Ctx { key: &self.0, enc })
+        }
+    }
+
+    impl AesXts256Ctx<'_> {
+        pub fn cipher(&self, tweak: &[u8; 16], data: &mut [u8]) -> Result<(), Error> {
+            // BCrypt only supports 64-bit tweaks, internally padding out the high 8
+            // bytes with zeroes. (Why?) This is fine for our purposes but it's a
```

Comment:
I don't know enough about the specific APIs at play here, but this certainly feels like an eyebrow raise

Reply:
Yes. But there's nothing we can do about it. They only recently added full 16-byte tweak support to symcrypt, and it does not appear that they have propagated it through bcrypt yet.

Perhaps ideally we'd just use symcrypt everywhere. But then we'd have to ship symcrypt.